### PR TITLE
Reference implementations for rsqrt and native_layer_norm

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds references for:
- `torch.rsqrt`
- `torch.native_layer_norm`
-  `torch.nn.functional.layer_norm`

`native_layer_norm` had a different number of dimensions if the input was 0-sized. I fixed that.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
It would be nice to have a way to display reference implementation code in documentation, this can compensate for unclear description (meanwhile it's being improved): https://github.com/pytorch/pytorch/issues/51455

**Comment by ngimel** [comment]
cc @chillee for native_layer_norm

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -1762,6 +1768,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
     return prims.slice_in_dim(a, start, start + length, axis=dim)
 
 
+def _normalize(a, norm_dims, eps):
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a_acc = prims.convert_element_type(a, computation_dtype)
+    biased_var, mean = var_mean(a_acc, dim=norm_dims, unbiased=False, keepdim=True)
```
Comment (ngimel) [inline torch/_refs/__init__.py:1774]: currently nvfuser doesn't understand var_mean, so this decomp is not readily usable

File: test/test_decomp.py [hunk 2]
```diff
@@ -450,6 +450,9 @@ def check_decomposed(aten_name):
         func = op.get_op()
         for sample_input in samples:
             if requires_grad:
+                if None in sample_input.args:
```
Comment (Chillee) [inline test/test_decomp.py:453]: Why is this needed?

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -1762,6 +1768,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
...
+    rstd = rsqrt(biased_var + eps)
```
Comment (Chillee) [inline torch/_refs/__init__.py:1775]: Can we just call `torch.rsqrt` here?

File: torch/_refs/__init__.py (hunk 1 above)
Comment (Chillee) [reply torch/_refs/__init__.py:1774]: Same here.

Also, the reason the previous decomp used separate calls to `var` and `mean` was that it more closely matched the numerics of eager mode. @ngimel's opinion was that the discrepancy shouldn't matter, but just noting in case you need to adjust tolerances on any tests.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -13445,6 +13519,23 @@ def error_inputs_mean(op_info, device, **kwargs):
...
+           supports_out=False,
+           assert_jit_shape_analysis=True,
+           sample_inputs_func=sample_inputs_native_layer_norm,
+           error_inputs_func=error_inputs_native_layer_norm,
+           supports_expanded_weight=True,
+           skips=(
+               # IndexError: tuple index out of range
+               DecorateInfo(unittest.skip('Skipped!'), 'TestGradients', 'test_forward_mode_AD'),
+               # Difference from float64 is larger with decomposition native_layer_norm_backward.default than original
+               DecorateInfo(unittest.skip('Skipped!'), 'TestDecomp', 'test_comprehensive',
```
Comment (Chillee) [inline torch/testing/_internal/common_methods_invocations.py:13536]: yeah actually, this might be what you're seeing.

File: torch/_refs/__init__.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1774]: I will make `mean` working with the nvFuser executor in a separate PR. `var` will be implemented later as it's currently a "prim" in PyTorch, but a composite in nvFuser.

It's separate calls to `var` and `mean`. Take a look at the `var_mean` function:
```
def var_mean(
    a: TensorLikeType,
    dim: Union[Optional[int], Optional[List[int]]] = None,
    unbiased: Optional[bool] = None,
    keepdim: bool = False,
    *,
    correction: Optional[int] = None,
):
    v = var(a, dim, unbiased, keepdim, correction=correction)
    m = mean(a, dim, keepdim)
    return v, m
```

File: torch/_refs/__init__.py (hunk 3 above)
Comment (Chillee) [reply torch/_refs/__init__.py:1775]: My understanding is that the current thinking is to prefer calling `torch.foo` vs. `_refs.foo` if convenient, as it's fairly easy to map from `torch.foo` => `_refs.foo` but not the other way around.

See @mruberry's comment here: https://github.com/pytorch/pytorch/pull/78689#discussion_r889159026

Comment (IvanYashchuk) [reply torch/_refs/__init__.py]: It's very easy to map `_refs.rsqrt` to `torch.rsqrt` because `_refs.rsqrt` is a wrapper of `torch.rsqrt` for regular and meta tensors and `test_decompy.py` passing indicates that it's fine as it is.

I changed this to `torch.rsqrt`.

File: torch/_refs/__init__.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1821]: PR for mean: https://github.com/pytorch/pytorch/pull/79444
PR for var: https://github.com/pytorch/pytorch/pull/79517
And together they make nvfuser understand var_mean.

**Comment by mruberry** [comment]
> It would be nice to have a way to display reference implementation code in documentation, this can compensate for unclear description (meanwhile it's being improved): #51455

https://github.com/pytorch/pytorch/issues/79518

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu [hunk 5]
```diff
@@ -852,21 +852,21 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_cuda(
   Tensor rstd = at::empty({M}, X->options().dtype(acc_type));
   if (M > 0) {
     LayerNormKernelImpl(*X, *gamma, *beta, M, N, eps, &Y, &mean, &rstd);
+  }
+  const auto input_shape = input.sizes();
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/layer_norm_kernel.cu:856]: Let's add a comment for this block -- is this a BC-breaking change?

Comment (IvanYashchuk) [reply aten/src/ATen/native/cuda/layer_norm_kernel.cu:856]: > is this a BC-breaking change?
Yes, it does change the shape of var and mean returned for input tensors with 0s in the shape. It's a bug-fix.

Comment (IvanYashchuk) [reply aten/src/ATen/native/cuda/layer_norm_kernel.cu:856]: Behavior on master:
```py
In [1]: import torch

In [2]: a = torch.randn(2, 0, 3, 3, device='cuda') # note 0 in the 2nd dim

In [3]: [x.shape for x in torch.native_layer_norm(a, (3, 3), None, None, 1e-5)]
Out[3]: [torch.Size([2, 0, 3, 3]), torch.Size([0]), torch.Size([0])] # shapes for var and mean are wrong

In [4]: a = torch.randn(2, 1, 3, 3, device='cuda') # note non-zero in the 2nd dim

In [5]: [x.shape for x in torch.native_layer_norm(a, (3, 3), None, None, 1e-5)]
Out[5]: [torch.Size([2, 1, 3, 3]), torch.Size([2, 1, 1, 1]), torch.Size([2, 1, 1, 1])]
```

File: torch/_refs/__init__.py (hunk 3 above)
Comment (Chillee) [reply torch/_refs/__init__.py]: From the perspective of the "correctness" tests it's fine. But, if I understand correctly, `_refs.rsqrt => prims.rsqrt`, and then you're saying that `prims.rsqrt` simply calls `torch.rsqrt` under the hood. 

But when we trace things out, using `_refs.rsqrt` doesn't have an interception point to allow us to decompose `native_layer_norm` but not decompose `_refs.rsqrt`.

TBH, for `rsqrt` this is a somewhat strange discussion haha since `_refs.rsqrt == prims.rsqrt == aten::rsqrt`. But it matters more for `var_mean`.

For example, say that I'm a compiler that wants `native_layer_norm` decomposed, but I have a special handling for `var_mean`. If we call `_refs.var_mean`, then our trace has no option to trace out `var_mean` - we'll get `var` followed by `mean`.

If we use `torch.var_mean` though, then we'll have the choice to either further decompose into `var` and `mean` or preserve `var_mean` in our trace.

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -1808,6 +1815,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
     return prims.slice_in_dim(a, start, start + length, axis=dim)
 
 
+def _normalize(a, norm_dims, eps):
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a_acc = prims.convert_element_type(a, computation_dtype)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1820]: Let's make this conversation conditional

File: torch/_refs/__init__.py [hunk 7]
```diff
@@ -1808,6 +1815,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
     return prims.slice_in_dim(a, start, start + length, axis=dim)
 
 
+def _normalize(a, norm_dims, eps):
```
Comment (mruberry) [inline torch/_refs/__init__.py:1818]: Add a comment for this function -- what it's meant to be used for, how it's intended to be used -- we should probably add type annotations to it

File: torch/_refs/__init__.py [hunk 8]
```diff
@@ -1808,6 +1815,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
...
+
+
+@register_decomposition(torch.ops.aten.native_layer_norm)
+def native_layer_norm(
+    input: TensorLikeType,
+    normalized_shape: ShapeType,
+    weight: Optional[Tensor],
+    bias: Optional[Tensor],
+    eps: float,
+) -> Tuple[TensorLikeType, TensorLikeType, TensorLikeType]:
```
Comment (mruberry) [inline torch/_refs/__init__.py:1834]: With the latest updates I think it's just all `Tensor` now

File: torch/_refs/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_refs/__init__.py]: Because of the way the current context mapping is implemented it's a one-way torch. -> refs. transform, so we should get in the habit of preferring torch. calls when possible. In some cases the torch operation doesn't have as much functionality as the ref, and in cases where that extra functionality is used the ref should be called explicitly, instead.

File: torch/_refs/__init__.py (hunk 1 above)
Comment (Lezcano) [reply torch/_refs/__init__.py:1827]: I believe @peterbell10 once told me that `var_mean` had some speed and stability problems in PyTorch (?). I do not remember very well so I'll let her discuss this, but perhaps this is relevant here.

Comment (peterbell10) [reply torch/_refs/__init__.py:1827]: On CPU, the mean from `var_mean` is less accurate than calling `mean` separately. `mean` is implemented roughly as `sum(x, dim) / x.size(dim)`, where `sum` uses a low-error summation algorithm. `var_mean` on the other hand computes both the mean and var in a single pass, but with a naive summation that is less accurate.

Comment (Chillee) [reply torch/_refs/__init__.py:1827]: For similar reasons, can we change this call to a `torch.var_mean` as well?

Comment (ngimel) [reply torch/_refs/__init__.py:1827]: Well, then, unless you decompose `torch.var_mean` further, you'd get inaccurate cpu result

**Comment by IvanYashchuk** [comment]
All CI is passing now. Since this PR adds a new OpInfo it revealed a few problems and tests had to be skipped. I submitted an issue https://github.com/pytorch/pytorch/issues/79705.

Except for `torch._refs.var_mean` vs `torch.var_mean` (https://github.com/pytorch/pytorch/pull/79413/files#r899336672) I addressed all the feedback.

@Chillee, @mruberry could you please take another look?

**Review by mruberry — APPROVED** [review]
Great work, @IvanYashchuk! Nice attention to the bug fix and the issue. 

Let's ship this as soon as the tests are happy.

**Review by Chillee — APPROVED** [review]
LGTM, besides the minor nit about `torch.var` and `torch.mean`.

I don't want to block this PR on that though - this PR has been through a bunch of rounds already :)

If it's an issue for using this decomposition in the future we can just modify it to `torch.var`/`torch.mean` then.
